Mathlib/Data/Finsupp/Weight.lean

```diff
@@ -41,68 +44,84 @@ there are finitely many `f : σ →₀ ℕ` of bounded weight.
 
 ## Degree
 
-- `Finsupp.degree`:  the weight when all components of `w` are equal to `1 : ℕ`.
-The present choice is to have it defined as a plain function.
+- `Finsupp.degree f` is the sum of all `f s`, for `s ∈ f.support`.
+The present choice is to have it defined as a function.
 
 - `Finsupp.degree_eq_zero_iff` says that `f.degree = 0` iff `f = 0`.
 
 - `Finsupp.le_degree` says that `f s ≤ f.degree`.
 
-- `Finsupp.degree_eq_weight_one` says `f.degree = f.weight 1`.
+- `Finsupp.degree_eq_weight_one` says `f.degree = f.weight 1` when `N` is a semiring.
 This is useful to access the additivity properties of `Finsupp.degree`
 
 - For `Finite σ`, `Finsupp.finite_of_degree_le` proves that
 there are finitely many `f : σ →₀ ℕ` of bounded degree.
 
 
-
 ## TODO
 
-* The relevant generality of these constructions is not clear.
-It could probably be generalized to `w : σ → M` and `f : σ →₀ N`,
-provided `N` is a commutative semiring and `M`is an `N`-module.
-
 * Maybe `Finsupp.weight w` and `Finsupp.degree` should have similar types,
 both `AddCommMonoidHom` or both functions.
 
 -/
 
-variable {σ M : Type*} (w : σ → M)
+variable {σ M N : Type*} [Semiring N] (w : σ → M)
 
 namespace Finsupp
 
 section AddCommMonoid
 
-variable [AddCommMonoid M]
-/-- The `weight` of the finitely supported function `f : σ →₀ ℕ`
-with respect to `w : σ → M` is the sum `∑ i, f i • w i`. -/
-noncomputable def weight : (σ →₀ ℕ) →+ M :=
-  (Finsupp.linearCombination ℕ w).toAddMonoidHom
+variable [AddCommMonoid M] [Module N M]
+/-- The `weight` of the finitely supported function `f : σ →₀ N`
+with respect to `w : σ → M` is the sum `∑(f i)•(w i)`. -/
+noncomputable def weight : (σ →₀ N) →+ M :=
+  (Finsupp.linearCombination N w).toAddMonoidHom
 
 @[deprecated weight (since := "2024-07-20")]
 alias _root_.MvPolynomial.weightedDegree := weight
 
-theorem weight_apply (f : σ →₀ ℕ) :
+theorem weight_apply (f : σ →₀ N) :
     weight w f = Finsupp.sum f (fun i c => c • w i) := rfl
 
 @[deprecated weight_apply (since := "2024-07-20")]
 alias _root_.MvPolynomial.weightedDegree_apply := weight_apply
 
+theorem weight_single_apply (s : σ) (c : M) (f : σ →₀ N) :
+    weight (single s c) f = f s • c := by
+  rw [weight_apply]
+  rw [Finsupp.sum_eq_single s]
+  · simp only [single_eq_same]
+  · intro i _ hi
+    rw [single_eq_of_ne hi.symm, smul_zero]
+  · intro _
+    simp only [single_eq_same, zero_smul]
```

Comment:
```suggestion
  rw [weight_apply, Finsupp.sum_single_index]
  simp -- or something like that
```
no?

Reply:
No…

Reply:
(The smul has to enter the single, and then it doesn't seem to reach to anywhere useful).